### DISCUSSION TITLE
Run some migrations, then run a seeder, and then run more migrations

### DISCUSSION THREAD

**Discussion by kasir-barati** [prisma/prisma#24959] [Q&A]
### Question

Hey Prisma folks,

I need to be able to write a couple of migrations and then run a seeder (preferably in TS/JS) and after that run more migrations. And I think this might happen in the future that I need to do it again. Consider this scenario:

1. I wanna add mandatory `birthdate` field to `users` table.
2. So I am creating a new migration script like this: `prisma migrate dev --name temp_birthdate_of_user --create-only`.
3. I have the data somewhere else which I can fetch over the network, or it is a simple snaphshot and I just need to restore it.
    - I am assuming that I can read it via TS/JS and then restore it via tools like [pg-dump-restore](https://www.npmjs.com/package/pg-dump-restore). 
    - Or maybe something a bit more rough like [this](https://stackoverflow.com/q/2732474/8784518).
4. Then I wanna create a new migration `prisma migrate dev --name birthdate_of_user --create-only`.
5. In this one I will use the restored data in order to fill the newly added `birthdate` field.
6. Afterward I wanna get rid of that temp database/table.
7. And now everything is done and we are good to go.

<hr />

#### Enquiries
 
I. How can  I do it in Prisma? 
II. Can I access my file system in migration files? 
III. I read [this Stackoverflow Q&A](https://stackoverflow.com/a/74455242/8784518) and I though if it is possible there, it might be here too. 

<hr />

### How to reproduce

I guess you could look at it here: https://github.com/kasir-barati/awesome-sql-orm

### Information about Prisma Schema, Client Queries and Environment

- OS: Manjaro.
- Database: PostgreSQL.
- Node.js version: v22.4.0

- `prisma -v`

```cmd
Environment variables loaded from .env
prisma                  : 5.17.0
@prisma/client          : 5.17.0
Computed binaryTarget   : debian-openssl-3.0.x
Operating System        : linux
Architecture            : x64
Node.js                 : v22.4.0
Query Engine (Node-API) : libquery-engine 393aa359c9ad4a4bb28630fb5613f9c281cde053 (at node_modules/.pnpm/@prisma+engines@5.17.0/node_modules/@prisma/engines/libquery_engine-debian-openssl-3.0.x.so.node)
Schema Engine           : schema-engine-cli 393aa359c9ad4a4bb28630fb5613f9c281cde053 (at node_modules/.pnpm/@prisma+engines@5.17.0/node_modules/@prisma/engines/schema-engine-debian-openssl-3.0.x)
Schema Wasm             : @prisma/prisma-schema-wasm 5.17.0-31.393aa359c9ad4a4bb28630fb5613f9c281cde053
Default Engines Hash    : 393aa359c9ad4a4bb28630fb5613f9c281cde053
Studio                  : 0.502.0
```

**Comment by ludralph** [comment]
Hi @kasir-barati 👋 

> I. How can I do it in Prisma?

Currently, Prisma does not natively support running migrations using TypeScript/JavaScript directly within the migration files. 

- JS/TS Migrations: There are multiple requests for supporting JavaScript/TypeScript in migrations to allow for more complex operations, such as interacting with third-party APIs or performing conditional logic. 
    - https://github.com/prisma/prisma/issues/20628
    - https://github.com/prisma/prisma/issues/6657

- Programmatic Access: There is also a request for programmatic access to Prisma Migrate CLI commands, which could potentially be used to script the migration process.
    - https://github.com/prisma/prisma/issues/4703

> II. Can I access my file system in migration files?

Prisma migrations are primarily SQL-based and do not support direct access to the file system within migration files.

However, you can achieve your goal by:

1. Create Initial Migration:
```
prisma migrate dev --name temp_birthdate_of_user --create-only
```
2. Edit the draft migration and add the new mandatory field in the table 
3. Write a script to seed the data

You may also want to look at the [Expand and Contract pattern](https://www.prisma.io/docs/orm/prisma-migrate/workflows/customizing-migrations#example-use-the-expand-and-contract-pattern-to-evolve-the-schema-without-downtime) for evolving your schema without downtime.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!

**Comment by spellitkit** [comment]
Just want to add my few cents on @kasir-barati question and describe the use case here.

tl;dr; I do think the problem is not feasibly solvable without versioning client types or writing untyped seeders, so the next best right now is to actually test data migration scripts manually. But I'm leaving the write down here if someone would want to tackle this for future ORMs.

The pattern for the issue here could be described as following:
- there is already an existing schema and extensive seeder script written for that schema
- schema migration is applied adding some new fields
- developer writes a new data migration that transfers data from old fields to new ones
- developer runs `pnpm prisma migrate dev` to apply migrations and seed the db
- developer finds out that his data migration did not modify any data because there was no data in the db - the seed script runs after migrations
- developer have to test the script manually, but there is no easy way for others to verify and tell if they actually did test that - it will never touch the actual seeded data in automatic way on local dev db.
- then the migration fails when executed in persistent deployment on actual data.

Generally I think there is no ideal solution here since seeder script uses the generated client that realistically only has latest schema types (after all migrations were run). So we can't run any theoretical intermediate seeders in between the migrations unless it's untyped scripts.

The issue is generally a dev workflow inconvenience for more mature apps. 

It would be really cool though to have a system that can track those intermediate seeder states that are designed to replicate each state of the persistent db between migrations.

**Reply by Ehbraheem** [reply]
I believe the most logical solution would be for future Prisma versions to support TypeScript/JavaScript-based migration files, similar to most mature ORMs. Data migration is a standard requirement in mature systems, and like schema migrations, it needs proper tracking for replayability and state management.

In frameworks like ActiveRecord, for example, migration files can contain Ruby scripts or raw SQL, allowing complex data manipulation as part of the migration lifecycle. This makes it possible to evolve both schema and data in a controlled, versioned way.

The current limitation in Prisma makes robust data migrations unnecessarily difficult. Most third-party or ad-hoc solutions attempt to fill this gap, but they often fall short because Prisma’s migration system is fundamentally schema-focused and does not provide the necessary primitives for tracking, versioning, and safely replaying data transformations. As a result, building reliable data migration workflows on top of it becomes fragile and inconsistent.

I hope the Prisma team prioritizes this important feature. As it stands, the limitation makes adoption harder for mature systems that require structured, versioned data migrations as part of their standard workflow.
